libraries/botbuilder-ai/src/qnaMakerDialog.ts

```diff
@@ -0,0 +1,359 @@
+    private hostName: any;
+    private endpointKey: any;
+    private threshold: number;
+    private top: number;
+    private noAnswer: Activity;
+    private activeLearningCardTitle: string;
+    private cardNoMatchText: string;
+    private strintFilters: any;
+    private cardNoMatchResponse: Activity;
+
+    /**
+     * Creates a new QnAMakerDialog instance.
+     * @param knowledgeBaseId The Id of the QnAMaker knowledgebase to be queried.
+     * @param endpointKey The endpoint key to use when querying the knowledgebase.
+     * @param hostName Hostname to be used to form the QnAMaker host URL, which follows the following format https://{hostName}.azurewebsites.net/qnamaker
+     * @param noAnswer (Optional) Activity to be sent in the event no answer is found within the knowledgebase.
+     * @param threshold (Optional) The threshold above which to treat answers found from the knowledgebase as a match.
+     * @param activeLearningCardTitle (Optional) Title of the card displayed showing active learning options if active learning is enabled.
+     * @param cardNoMatchText (Optional) Text to be show on a button alongside active learning options, allowing a user to indicate none of the options are applicable.
+     * @param top (Optional) Maximum number of answers to return from the knowledgebase.
+     * @param cardNoMatchResponse (Optional) Activity to be sent if the user selects the no match option on an active learning card.
+     * @param strictFilters (Optional) QnAMakerMetadata collection used to filter / boost queries to the knowledgebase.
+     * @param dialogId (Optional) Id of the created dialog. Default is 'QnAMakerDialog'.
+     */
+    constructor(knowledgeBaseId, endpointKey, hostName, noAnswer = null, threshold = 0.3, activeLearningCardTitle = 'Did you mean:', cardNoMatchText = 'None of the above.', top = 3, cardNoMatchResponse: Activity = null, strictFilters: QnAMakerMetadata[] = null, dialogId = 'QnAMakerDialog') {
+        super(dialogId);
+
+        this.knowledgeBaseId = knowledgeBaseId;
+        this.hostName = hostName;
+        this.endpointKey = endpointKey;
+        this.threshold = threshold;
+        this.top = top;
+        this.activeLearningCardTitle = activeLearningCardTitle;
+        this.cardNoMatchText = cardNoMatchText;
+        this.strintFilters = strictFilters;
+        this.noAnswer = noAnswer;
+        this.cardNoMatchResponse = cardNoMatchResponse;
+
+        this.addStep(this.callGenerateAnswer.bind(this));
+        this.addStep(this.callTrain.bind(this));
+        this.addStep(this.checkForMultiTurnPrompt.bind(this));
+        this.addStep(this.displayQnAResult.bind(this));
+    }
+
+    public async beginDialog(dc: DialogContext, options?: object): Promise<DialogTurnResult> {
+        if (!dc) { throw new Error('Missing DialogContext'); }
+        
+        if (dc.context.activity.type != ActivityTypes.Message) 
+        { 
+            return Dialog.EndOfTurn; 
+        }
+
```

Comment:
The problem with this is that the waterfalls state won't be properly initialized since you don't call super.beginDialog().

If you're doing this because you could be a root dialog then I would call endDialog() instead. If you're not a root dialog, it's very unlikely you'd get started by a non-message activity.

Reply:
I have switched to use endDialog()